pytorch_lightning/strategies/collaborative.py

```diff
@@ -0,0 +1,511 @@
+                        "We tried to infer the batch size from the first batch of data are were unable to. "
+                        "Please provide the batch size to the Strategy (CollaborativeStrategy"
+                        f"exception raised: {e}"
+                    )
+            self._initialize_hivemind()
+
+    @property
+    def num_peers(self) -> int:
+        if self.opt:
+            return self.opt.tracker.global_progress.num_peers
+        return 1
+
+    @property
+    def dht(self) -> "hivemind.DHT":
+        return self.dht_manager.dht
+
+    @property
+    def on_tpu(self) -> bool:
+        return self.root_device.type == "xla" and _XLA_AVAILABLE
+
+    @property
+    def on_gpu(self) -> bool:
+        return self.root_device.type == "cuda" and torch.cuda.is_available()
+
+    def reduce(self, tensor: Union[Any, torch.Tensor], *args: Any, **kwargs: Any) -> Union[Any, torch.Tensor]:
+        return tensor
+
+    def all_gather(self, tensor: torch.Tensor, group: Optional[Any] = None, sync_grads: bool = False) -> torch.Tensor:
+        return tensor
+
+    @property
+    def root_device(self) -> torch.device:
+        # todo: cyclic import, someone show me the way
+        from pytorch_lightning.accelerators.cpu import CPUAccelerator
+        from pytorch_lightning.accelerators.gpu import GPUAccelerator
+
+        if isinstance(self.accelerator, GPUAccelerator):
+            return torch.device(f"cuda:{torch.cuda.current_device()}")
+        elif isinstance(self.accelerator, CPUAccelerator):
+            return torch.device("cpu")
+        raise ValueError
+
+    def model_to_device(self) -> None:
+        assert self.model is not None
+        self.model.to(self.root_device)
+
+    def setup(self, trainer: "pl.Trainer") -> None:
+        self.model_to_device()
+        super().setup(trainer)
+        if self.precision_plugin.precision in (PrecisionType.HALF, PrecisionType.MIXED):
+            self.precision_plugin.scaler = hivemind.GradScaler()
+
+    @property
+    def is_global_zero(self) -> bool:
+        return True
+
+    def barrier(self, *args: Any, **kwargs: Any) -> None:
+        pass
+
+    def broadcast(self, obj: TBroadcast, src: int = 0) -> TBroadcast:
+        return obj
+
+    def teardown(self) -> None:
+        super().teardown()
+        if self.on_gpu and self.lightning_module is not None:
+            # GPU teardown
+            self.lightning_module.cpu()
+            # clean up memory
+            torch.cuda.empty_cache()
+        if self.opt:
+            self.opt.shutdown()
+        log.info("Shutting down hivemind DHT.")
+        self.dht.shutdown()
+
+
+class HiveMindScheduler:
+    """Wrapper for schedulers to prevent Lightning from stepping the scheduler too soon.
+
+    This code ensures that we only step when the HiveMind optimizer reaches the global step.
+    """
+
+    def __init__(self, optimizer: "hivemind.Optimizer", scheduler: LRSchedulerTypeUnion):
```

Comment:
Someone pls help me figure out this MyPy issue 😢 

```
pytorch_lightning/strategies/collaborative.py:229: error: Argument "scheduler" to "HiveMindScheduler" has incompatible type "Union[_LRScheduler, ReduceLROnPlateau]"; expected "Union[_LRScheduler, ReduceLROnPlateau]"  [arg-type]
```

I've tried various configurations from the Types utilities but have had no success.